File: 53-maximum-subarray/maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int n=nums.length;
        int ans=0;
        int s=0;
        int maxi=-100000;
        for(int i=0;i<n;i++)
            maxi=Math.max(maxi,nums[i]);
        if(maxi<0)
            return maxi;
        for(int i=0;i<n;i++)
        {
            s=s+nums[i];
            if(s<0)
                s=0;
            else{
                ans=Math.max(ans,s);
                System.out.println(ans);
            }
        }
        return ans;
        
    }
}

File: 53-maximum-subarray/maximum-subarray.cpp
//This is KADANE , works for atleast 1 +ve ele in array
// class Solution {
// public:
//     int maxSubArray(vector<int>& nums) {
//        int len=nums.size();
//         int sum=0;
//         int ans=INT_MIN;
//         for(int i=0;i<len;i++)
//         {
//             if((sum+nums[i])>=0)
//             {
//                 sum+=nums[i];
//                 ans=max(ans,sum);
//             }
//             else
//             {
//                 sum=0;
//             }
//         }
//         return ans;
//     }
// };


class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n=nums.size();
        int mini=INT_MAX;
        for(auto i: nums)
        {
            if(i<=0)
                mini=min(mini,abs(i));
        }
        cout<<mini<<endl;
        int i=0,j=0;
        int s=0;int ms=-1*mini;
        while(j<n)
        {
            cout<<s<<endl;
            s=s+nums[j];
            if(s<=0)
            {
                j++;
                i=j;
                s=0;
                continue;
            }
            ms=max(ms,s);
            j++;
        }
        return ms;
    }
};
/**

   int len=nums.size();
        int sum=0;
        int mss=nums[0]; //max sum difference
        int mini=0; //lowest sum
        for(int i=0;i<len;i++)
        {
            sum+=nums[i];
            mss=max(mss,sum-mini);
            mini=min(sum,mini);
        }
        return mss;

*/